client/src/main/scala/drt/client/services/SPACircuit.scala

```diff
@@ -601,48 +632,99 @@ class ShiftsHandler[M](modelRW: ModelRW[M, Pot[String]]) extends LoggingActionHa
+
+    val lines = rawFixedPoints.split("\n").toList.map(line => {
+      val withoutTerminal = line.split(",").toList.map(_.trim)
+      val withTerminal = withoutTerminal match {
+        case fpName :: tail => fpName.toString :: terminalName :: todayString :: tail
+        case _ => Nil
+      }
+      withTerminal.mkString(", ")
+    })
+    lines.mkString("\n")
+  }
 }
 
-class StaffMovementsHandler[M](modelRW: ModelRW[M, Seq[StaffMovement]]) extends LoggingActionHandler(modelRW) {
+class FixedPointsHandler[M](modelRW: ModelRW[M, Pot[String]]) extends LoggingActionHandler(modelRW) {
+  protected def handle = {
+    case SetFixedPoints(fixedPoints: String, terminalName: Option[String]) =>
+      if (terminalName.isDefined)
+        updated(Ready(fixedPoints), Effect(Future(RunTerminalSimulations(terminalName.get))))
+      else
+        updated(Ready(fixedPoints), Effect(Future(RunAllSimulations())))
+    case SaveFixedPoints(fixedPoints: String, terminalName: TerminalName) =>
+      val otherTerminalFixedPoints = FixedPoints.filterOtherTerminals(terminalName, value.getOrElse(""))
+      val newRawFixedPoints = otherTerminalFixedPoints + "\n" + fixedPoints
+      AjaxClient[Api].saveFixedPoints(newRawFixedPoints).call()
+      effectOnly(Effect(Future(SetFixedPoints(newRawFixedPoints, Option(terminalName)))))
+    case AddShift(fixedPoints) =>
+      updated(Ready(s"${value.getOrElse("")}\n${fixedPoints.toCsv}"))
+    case GetFixedPoints() =>
+      effectOnly(Effect(AjaxClient[Api].getFixedPoints().call().map(res => SetFixedPoints(res, None))))
+  }
+}
+
+class StaffMovementsHandler[M](modelRW: ModelRW[M, Pot[Seq[StaffMovement]]]) extends LoggingActionHandler(modelRW) {
   protected def handle: PartialFunction[Any, ActionResult[M]] = {
     case AddStaffMovement(staffMovement) =>
-      val v: Seq[StaffMovement] = value
-      val updatedValue: Seq[StaffMovement] = (v :+ staffMovement).sortBy(_.time)
-      updated(updatedValue)
+      if (value.isReady) {
+        val v: Seq[StaffMovement] = value.get
+        val updatedValue: Seq[StaffMovement] = (v :+ staffMovement).sortBy(_.time)
+        updated(Ready(updatedValue))
+      } else noChange
     case RemoveStaffMovement(idx, uUID) =>
-      val updatedValue = value.filter(_.uUID != uUID)
-      updated(updatedValue, Effect(Future(SaveStaffMovements())))
+      if (value.isReady) {
+        val terminalAffectedOption = value.get.find(_.uUID == uUID).map(_.terminalName)
+        if (terminalAffectedOption.isDefined) {
+          val updatedValue = value.get.filter(_.uUID != uUID)
+          updated(Ready(updatedValue), Effect(Future(SaveStaffMovements(terminalAffectedOption.get))))
+        } else noChange
+      } else noChange
     case SetStaffMovements(staffMovements: Seq[StaffMovement]) =>
-      updated(staffMovements, Effect(Future(RunAllSimulations())).after(30 seconds))
+      updated(Ready(staffMovements), Effect(Future(RunAllSimulations())) /*.after(30 seconds)*/)
```

Comment:
Does this work initially without the 30 second poll? Bare in mind that it actually needs to wait for a bunch of other things to be ready in addition to the staff movements before it will work.